dhis2/dhis2-core
# refactor: AttributeValues as immutable, memory efficient value object [DHIS2-17482]

### Summary
Replaces `Set<AttributeValue>` with a new abstract type `AttributeValues` to reduce the memory footprint.

It became clear that having `Attribute` referenced in `AttributeValue` also had high complexity costs. Plenty of code was dealing with converting between shallow and full `Attribute` objects within `AttributeValue` objects. To simplify `AttributeValue` got removed in favor of using its ID in APIs. 

For backwards compatibility the API still allows to query further `Attribute` properties of attribute values using `fields=attributeValues[id,name]`. These are added into the response in the field filtering not much different than before where the shallow `Attribute` would be replaced by the full object.

### Implementation Notes
**Decoupling**: `AttributeValues` was solved as an interface to avoid usage to be bound again to a specific implementation.

**Laziness**: The `LazyAttributeValues` implementation takes the JSON string from DB and wraps it as such. 
It is lazy because the JSON from which it is constructed is only parsed once a property is accessed. 
This means when attribute values are only fetched because they are eagerly mapped but never actually used the costs are minimal.

**Compactness**: The map of attribute ID and values is represented as a `String[]` for the attribute IDs and a `String[]` for the values of the exact length required. This assures good data locality but most of all is very memory efficient.
In addition all attribute ID strings are known to have a high potential to reoccur (e.g. same attribute in all OU) so they share instances using `String#intern()`. 

**Lookup**: `attributeValues` in identifiable objects no longer require "caching" for access as the array of IDs is always sorted whereby a binary search _O(log n)_ can be used for lookup in case many attributes are in a set. This also further contributes to a small memory footprint.

**Immutability**: `AttributeValues` are immutable snapshots preventing unintentional modification. Internally, however, they do release memory if possible. Once the JSON origin value gets parsed it is unassigned to allow GC, again to keep footprint low.
As API and DB do **not** share the same JSON representation keeping the JSON also rarely is beneficial.   

### Compatibility Note
I found some tests that would use attribute values in JSON other than a JSON string. This was never correct but jackson apparently did allow it and mapped numbers and booleans to string equivalent. The new code will require a valid JSON string. I updated the test data accordingly assuming that API users should also just assume that JSON strings are valid.

### Cleanup
Going through the places attributes and attribute values were involved in I cleaned up everything that seemed unclear without longer analysis. 

- [x] Use SQL in attribute related queries (also to use proper functions that should have best DB query performance)
- [x] attribute value validation got greatly simplified removing abstraction that were used to organize the code (a switch and a few methods was all that is needed)
- [x] several methods from the `GenericStore` and `IdentifiableObjectManager` could be removed as they were used (this affects mostly methods targeting attributes but also a few others)
- [x] methods in `GenericStore` and `IdentifiableObjectManager` that deal with attributes accept their ID as parameter, not objects (in stores I always used `UID` to ensure better protection against SQL injection)
- [x] add `tableName` field to stores (needed for generic SQL) 
- [x] remove all user access and user group access stores, these where unused but also these were wrongly declared stored as they never had a correspondence in an object that is mapped to a table (they are just embedded objects in a field/column)
- [x] numerous small refactoring highlighted and executed automatically by the IDE (just stuff that is safe)

### Automatic Testsing
New suit of unit tests were added for the `AttributeValues` implementation.

Many existing tests needed adjustment in their setup and validation of attribute values.
Deleted test cases did target otherwise unused methods that got removed from the API.

### Manual Testing
There are many things one could test but a minimal test would be 

* create an attribute, allow it for some object type that is easy to create
* create an object with that type and set the attribute value
* check the attribute value got stored
* update the attribute value for the object and check the change is effective

### Next
- [ ] In the old field filtering the attribute handling got removed, meaning you can no longer successfully request the equivalent of  `fields=attributeValues[attribute[x]]` where `x` is any other `Attribute` property than `id`. However, this does not affect the metadata API as that uses the new field filtering where the feature remains. A basic analysis shows that the old field filtering is mostly used in test code. We should replace it and remove the old field filtering code.
